lib/pgBackRest/Storage/S3/Auth.pm

```diff
@@ -0,0 +1,262 @@
+use constant S3                                                     => 's3';
+use constant AWS4                                                   => 'AWS4';
+use constant AWS4_REQUEST                                           => 'aws4_request';
+use constant AWS4_HMAC_SHA256                                       => 'AWS4-HMAC-SHA256';
+
+use constant S3_HEADER_AUTHORIZATION                                => 'authorization';
+    push @EXPORT, qw(S3_HEADER_AUTHORIZATION);
+use constant S3_HEADER_DATE                                         => 'x-amz-date';
+    push @EXPORT, qw(S3_HEADER_DATE);
+use constant S3_HEADER_CONTENT_SHA256                               => 'x-amz-content-sha256';
+    push @EXPORT, qw(S3_HEADER_CONTENT_SHA256);
+use constant S3_HEADER_HOST                                         => 'host';
+    push @EXPORT, qw(S3_HEADER_HOST);
+
+use constant PAYLOAD_DEFAULT_HASH                                   => sha256_hex('');
+    push @EXPORT, qw(PAYLOAD_DEFAULT_HASH);
+
+####################################################################################################################################
+# s3DateTime - format date/time for authentication
+####################################################################################################################################
+sub s3DateTime
+{
+    # Assign function parameters, defaults, and log debug info
+    my
+    (
+        $strOperation,
+        $lTime,
+    ) =
+        logDebugParam
+        (
+            __PACKAGE__ . '::s3DateTime', \@_,
+            {name => 'lTime', default => time(), trace => true},
+        );
+
+    # Return from function and log return values if any
+    return logDebugReturn
+    (
+        $strOperation,
+        {name => 'strDateTime', value => strftime("%Y%m%dT%k%M%SZ", gmtime($lTime)), trace => true}
+    );
+}
+
+push @EXPORT, qw(s3DateTime);
+
+####################################################################################################################################
+# s3CanonicalRequest - strictly formatted version of the HTTP request used for signing
+####################################################################################################################################
+sub s3CanonicalRequest
+{
+    # Assign function parameters, defaults, and log debug info
+    my
+    (
+        $strOperation,
+        $strVerb,
+        $strUri,
+        $strQuery,
+        $hHeader,
+        $strPayloadHash,
+    ) =
+        logDebugParam
+        (
+            __PACKAGE__ . '::s3CanonicalRequest', \@_,
+            {name => 'strVerb', trace => true},
+            {name => 'strUri', trace => true},
+            {name => 'strQuery', trace => true},
+            {name => 'hHeader', trace => true},
+            {name => 'strPayloadHash', trace => true},
+        );
+
+    # Create the canonical request
+    my $strCanonicalRequest =
+        "${strVerb}\n${strUri}\n${strQuery}\n";
+    my $strSignedHeaders;
```

Comment:
Unusual, I know. I was trying to stay as close as possible the AWS terminology to make this code easier to parse/update later.